soco/discovery.py

```diff
@@ -215,27 +242,240 @@ def any_soco():
             d for d in cls._instances[cls._class_group].values() if d.is_visible
         )
     except (KeyError, StopIteration):
-        devices = discover()
+        devices = discover(allow_network_scan=allow_network_scan, **network_scan_kwargs)
         return None if devices is None else devices.pop()
 
     return device
 
 
-def by_name(name):
+def by_name(name, allow_network_scan=False, **network_scan_kwargs):
     """Return a device by name.
 
     Args:
         name (str): The name of the device to return.
+        allow_network_scan (bool, optional): If normal discovery fails, fall
+            back to a scan of the attached network(s) to detect Sonos
+            devices.
+        **network_scan_kwargs: Arguments for the `scan_network` function.
+            See its docstring for details.
 
     Returns:
-        :class:`~.SoCo`: The first device encountered among all zone with the
+        :class:`~.SoCo`: The first device encountered among all zones with the
         given player name. If none are found `None` is returned.
     """
-    devices = discover()
+    devices = discover(allow_network_scan=allow_network_scan, **network_scan_kwargs)
     if devices is None:
         return None
 
     for device in devices:
         if device.player_name == name:
             return device
     return None
+
+
+def _is_ipv4_address(ip_address):
+    """Helper function to test for an IPv4 address.
+
+    Args:
+        ip_address (str): The IP address to be tested, e.g.,
+            "192.168.1.35".
+
+    Returns:
+        bool: True if this is a well-formed IPv4 address.
+    """
+    try:
+        ipaddress.IPv4Network(ip_address)
+        return True
+    except ValueError:
+        return False
+
+
+def _find_ipv4_networks(min_netmask):
+    """Discover attached IP networks.
+
+    Helper function to return a set of IPv4 networks to which
+    the network interfaces on this node are attached.
+    Excludes public and loopback network ranges.
+
+    Args:
+        min_netmask(int): The minimum netmask to be used.
+
+    Returns:
+        set: A set of `ipaddress.ip_network` instances.
+    """
+
+    ipv4_net_list = set()
+    adapters = ifaddr.get_adapters()
+    for adapter in adapters:
```

Comment:
Here's a simpler variant (to avoid multi-level nesting), if you wish:
```
# Get list of IPv4 networks
nets = [ip_network(f"{ip.ip}/{ip.network_prefix}", strict=False)   # strict=False allows host bits in the address
        for adapter in ifaddr.get_adapters()
        for ip in adapter.ips 
        if not isinstance(ip.ip, tuple)  # IPv6 addresses are reported as tuples by ifaddr
]
# Filter out loopback and public networks
nets = [net for net in nets if net.is_private and not net.is_loopback]

# unfortunately there is no easy way to get a network using given prefixlen without touching internals..
def _minimize_netmask(net, min):
    if net.prefixlen < min:
        net._prefixlen = min
    return net

# Get networks with the minimum netmask
nets = [_minimize_netmask(net, min_netmask) for net in nets]
```

Reply:
I've kept the nested form for now.